lib/models/package-info-cache.js

```diff
@@ -0,0 +1,810 @@
+ */
+function getRealFilePath(file) {
+  let realPath;
+  
+  try {
+    realPath = realFilePathCache[file];
+
+    if (realPath) {
+      return realPath;
+    }
+
+    let stat = fs.statSync(file);
+    
+    if (stat.isFile() || stat.isFIFO()) {
+      realPath = fs.realpathSync(file);
+    }
+  } catch (e) {
+    if (e !== null && (typeof e === 'object') && (e.code === "ENOENT" || e.code === "ENOTDIR")) {
+      realPath = null;
+    } else {
+      throw e;
+    }
+  }
+  
+  realFilePathCache[file] = realPath;
+  return realPath;
+}
+
+/**
+ * Resolve the real path for a directory, return null if does not
+ * exist or is not a directory, return the real path otherwise.
+ *
+ * @private
+ * @method getRealDirectoryPath
+ * @param  {String} file file path
+ * @return {String} real path or null
+ */
+function getRealDirectoryPath(dir) {
+  let realPath;
+  
+  try {
+    realPath = realDirectoryPathCache[dir];
+
+    if (realPath) {
+      return realPath;
+    }
+
+    let stat = fs.statSync(dir);
+
+    if (stat.isDirectory()) {
+      realPath = fs.realpathSync(dir);
+    }
+  } catch (e) {
+    if (e !== null && (typeof e === 'object') && (e.code === "ENOENT" || e.code === "ENOTDIR")) {
+      realPath = null;
+    } else {
+      throw e;
+    }
+  }
+
+  realDirectoryPathCache[dir] = realPath;
+  return realPath;
+}
+
+/**
+ * Small utility class to store a single error during processing.
+ */
+class ErrorEntry {
+  constructor(type, data) {
+    this.type = type;
+    this.data = data;
+  }
+}
+
+/**
+ * Small utility class to store a list of errors during processing.
+ * Instances of this exist in the PackageInfo and PackageInfoCache
+ * objects.
+ */
+class ErrorList {
+  constructor() {
+    this.errors = [];
+  }
+
+  /**
+   * Add an error. The error obj is optional, and can be anything.
+   * We do this so we don't really need to create a series of error
+   * classes.
+   */
+  addError(errorType, errorData) {
+    this.errors.push(new ErrorEntry(errorType, errorData));
+  }
+
+  insertError(errorEntry) {
+    this.errors.push(errorEntry);
+  }
+
+  getErrors() {
+    return this.errors;
+  }
+
+  hasErrors() {
+    return (this.errors.length > 0);
+  }
+}
+
+const PACKAGE_JSON = "package.json";
+
+const ERROR_PACKAGE_DIR_MISSING = "packageDirectoryMissing";
+const ERROR_PACKAGE_JSON_MISSING = "packageJsonMissing";
+const ERROR_PACKAGE_JSON_PARSE = "packageJsonParse";
+const ERROR_EMBER_ADDON_MAIN_MISSING = "emberAddonMainMissing";
+const ERROR_DEPENDENCIES_MISSING = "dependenciesMissing";
+const ERROR_NODEMODULES_ENTRY_MISSING = "modulesEntryMissing";
+
+class PackageInfo {
+  constructor(pkgObj, realPath, cache) {
+    this.pkg = pkgObj;
+    this.pkg['ember-addon'] = this.pkg['ember-addon'] || {};
+    this.realPath = realPath;
+    this.cache = cache;
+    this.errors = new ErrorList();
+
+    // other fields that will be set as needed:
+    //  this.addonConstructor (ctor function or object - addons only)
```

Comment:
we should likely set then as `undefined` or `null` in the constructor, so the JIT can shape these objects all the same.

Reply:
Nice idea! Doing that now.